pkg/ipamd/datastore/data_store.go

```diff
@@ -1214,10 +1192,103 @@ func (ds *DataStore) GetFreePrefixes() int {
 
 func (ds *DataStore) CleanupDataStore(eniID string, ipv4 string, force bool, enableIpv4PrefixDelegation bool) error {
 	var err error
+	var ipv4Cidr net.IPNet
+	var ipv4CidrPtr *net.IPNet
 	if !enableIpv4PrefixDelegation {
-		err = ds.DelIPv4AddressFromStore(eniID, ipv4, force)
+		ipv4Cidr = net.IPNet{IP: net.ParseIP(ipv4), Mask: net.IPv4Mask(255, 255, 255, 255)}
 	} else {
-		err = ds.DelIPv4PrefixFromStore(eniID, ipv4, force)
+		_, ipv4CidrPtr, err = net.ParseCIDR(ipv4)
+		if err != nil {
+			return err
+		}
+		ipv4Cidr = *ipv4CidrPtr
 	}
+	err = ds.DelIPv4CidrFromStore(eniID, ipv4Cidr, force)
 	return err
 }
+
+func getFreeIPv4AddrfromCidr(availableCidr *AssignedIPv4Addresses) (string, error) {
+	if availableCidr == nil {
+		log.Errorf("Prefix datastore not initialized")
+		return "", errors.New("Prefix datastore not initialized")
+	}
+	strPrivateIPv4, err := getUnusedIP(availableCidr)
+	if err != nil {
+		log.Debugf("Get free IP from prefix failed %v", err)
+		return "", err
+	}
+	availableCidr.UsedIPs++
+	log.Debugf("Returning Free IP %s", strPrivateIPv4)
+	return strPrivateIPv4, nil
+}
+
+/*
+  /28 -> 10.1.1.1/32[out of cooldown], 10.1.1.2/32[out of cooldown], 10.1.1.3/32[assigned]
+  cached ip = 10.1.1.1/32
+  /28 ->  10.1.1.3/32[assigned]
+  return 10.1.1.1/32
+*/
+
+func getUnusedIP(availableCidr *AssignedIPv4Addresses) (string, error) {
+	//Check if there is any IP out of cooldown
+	var cachedIP string
+	for _, addr := range availableCidr.IPv4Addresses {
+		if !addr.Assigned() && !addr.inCoolingPeriod() {
+			//if the IP is out of cooldown and not assigned then cache the first available IP
+			//continue cleaning up the DB, this is to avoid stale entries and a new thread :)
+			if cachedIP == "" {
+				cachedIP = addr.Address
+			}
+			delete(availableCidr.IPv4Addresses, addr.Address)
+		}
+	}
+
+	if cachedIP != "" {
+		return cachedIP, nil
+	}
+
+	//If not in cooldown then generate next IP
+	ip, ipnet, err := net.ParseCIDR(availableCidr.Cidr.String())
+	if err != nil {
+		return "", err
+	}
+
+	for ip := ip.Mask(ipnet.Mask); ipnet.Contains(ip); getNextIPv4Addr(ip) {
```

Comment:
We might end up looping through a particular prefix twice here in bunch of cases. Can't we merge both the loops and check `isAssgined` & `inCoolingPeriod` flags on an IP address as we're traversing the prefix? If there is a strong case, to maintain two for loops we can maybe optimize it by checking on `UsedIP` count after the first traversal before trying to walk through the prefix again. For example, if `UsedIP` count is 16 after the first traversal we can return right away.

Reply:
Yeah, the first loop is to avoid adding a cleanup thread to release the IPs. Yes, I can do that, before I used to check for FreeIPs, during this refactoring it got missed, will add it here - 
```
if availableCidr.IsPrefix &&   availableCidr.usedIPs != 16 {					
    strPrivateIPv4, err := ds.getFreeIPv4AddrfromCidr(availableCidr)
```

Reply:
Yeah makes sense :)